app/uk/gov/hmrc/gform/controllers/FormController.scala

```diff
@@ -105,11 +107,36 @@ class FormController @Inject() (
     val fieldData = getFormData(cache.form)
     val envelopeF = fileUploadService.getEnvelope(cache.form.envelopeId)
 
+    //    val validationF: Future[ValidatedType] = validateSections.flatMap { t =>
+    //      validatedDataResultF.map { x =>
+    //        val y: Either[GformError, Unit] = for {
+    //          _ <- x.toEither
+    //          _ <- t.toEither
+    //        } yield ()
+    //        Validated.fromEither(y)
+    //      }
+    //    }
     def getErrors(sections: List[Section], data: Map[FieldId, Seq[String]], envelope: Envelope, envelopeId: EnvelopeId) = {
       Logger.debug(data + "this is data in get errors")
       val fields = repeatService.atomicFields(sections(sectionNumber.value))
       val allFields = sections.flatMap(repeatService.atomicFields)
-      Future.sequence(fields.map(fv => validationService.validateComponents(fv, data, envelopeId))).map(Monoid[ValidatedType].combineAll).map { validationResult =>
+      val x: Future[Either[GformError, Unit]] = Future.sequence(fields.map(fv => validationService.validateComponents(fv, data, envelopeId))).map(Monoid[ValidatedType].combineAll).map(_.toEither)
+      val y = validationService.validateSections(sections(sectionNumber.value), data, cache.form.envelopeId)(_.validate(data)(x =>
+        gformConnector.validatePostCodeUtr(x._1, x._2))).map(_.toEither)
+
+      val z: Future[ValidatedType] = {
```

Comment:
Let's refactor it together. I've tried to read and understand it but after 5 minutes got timeout.
I thinkg that's time to _transform monads_ :)

Reply:
I am all for trying out some monads. It is a confusing function. I expected some refactoring.